flink-tests/src/test/java/org/apache/flink/test/checkpointing/RestoreUpgradedJobITCase.java

```diff
@@ -0,0 +1,419 @@
+            savepointPath =
+                    jobClient
+                            .stopWithSavepoint(
+                                    true,
+                                    temporaryFolder.getRoot().getAbsolutePath(),
+                                    SavepointFormatType.NATIVE)
+                            .get();
+        } else {
+            throw new IllegalArgumentException("Unknown checkpoint type: " + checkpointType);
+        }
+        return savepointPath;
+    }
+
+    private static class IntSink implements SinkFunction<Integer> {
+        private final SharedReference<AtomicLong> result;
+
+        public IntSink(SharedReference<AtomicLong> result) {
+            this.result = result;
+        }
+
+        @Override
+        public void invoke(Integer value, Context context) throws Exception {
+            result.get().addAndGet(value);
+        }
+    }
+
+    private static class StringSink implements SinkFunction<String> {
+        private final SharedReference<AtomicLong> result;
+
+        public StringSink(SharedReference<AtomicLong> result) {
+            this.result = result;
+        }
+
+        @Override
+        public void invoke(String value, Context context) throws Exception {
+            result.get().addAndGet(Integer.parseInt(value));
+        }
+    }
+
+    private static class IntMap extends AbstractMap<Integer> {
+
+        private IntMap(int id) {
+            super(id);
+        }
+
+        @Override
+        public Integer map(Integer value) throws Exception {
+            return calculate(value);
+        }
+    }
+
+    private static class StringMap extends AbstractMap<String> {
+
+        private StringMap(int id) {
+            super(id);
+        }
+
+        @Override
+        public String map(String value) throws Exception {
+            return String.valueOf(calculate(Integer.parseInt(value)));
+        }
+    }
+
+    private abstract static class AbstractMap<T extends Object> extends RichMapFunction<T, T>
+            implements CheckpointedFunction {
+        private ListState<Integer> valueState;
+        private final int id;
+        private int state;
+
+        private AbstractMap(int id) {
+            this.id = id;
+        }
+
+        protected int calculate(int value) throws Exception {
+            return value + state;
+        }
+
+        @Override
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
+            valueState.add(id);
+        }
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {
+            this.valueState =
+                    context.getOperatorStateStore()
+                            .getListState(new ListStateDescriptor<>("state", Types.INT));
+            Iterator<Integer> iterator = valueState.get().iterator();
+            if (iterator.hasNext()) {
+                state = id * iterator.next();
+            }
```

Comment:
Yes, sure. we can use `checkState(id == iterator.next);`. Initially, I just wanted to avoid any `asserts` in operators since it is not transparent. But it would be more understandable rather than the final number result. I need to think